### ISSUE TITLE
salesforce: Bulk V2 api support for `bulk()` data import function

### ISSUE THREAD

**Issue by aleksa-krolls** [OpenFn/adaptors#407]
## Adaptor
salesforce

## Background
Recently in some Salesforce projects, when we're connecting to sandboxes that are in the latest version of Salesforce... we're finding that bulk import jobs aren't closing when using our adaptors `bulk()` function that connects to the Bulk API V1. We're seeing way more timeouts than we're used to, so we want to upgrade our adaptor to work with the Bulk API V2 to see if that improves job performance. 
<img width="1220" alt="Screen Shot 2023-10-10 at 10 51 33 PM" src="https://github.com/OpenFn/adaptors/assets/16758106/a640a3df-35b0-4170-9fd5-d0f59b862b7c">
^these jobs remain "in progress" for a very long time and take forever to process, and then our OpenFn runs [timeout](https://www.openfn.org/projects/wfwi-committed-giving-staging/runs/06525aa1-78e9-7a57-800a-f79219183b48). 

## Request
Please develop a new `bulk()` function that leverages the Salesforce Bulk V2 api. 
- Users should be able to upsert/insert/update records in bulk with this function.
- I don't think JSForce supports this new API version, so I think we will need to chain a series of http requests to deliver this new function (similar to how we implemented the `bulkQuery()` function. 
- Consider that we do have several existing implementations that use the Bulk API V1 function we currently have, so ideally this new version behaves just like the existing `bulk()` function and has the same options. The only main difference is that we should be leveraging the Bulk V2 api. 

See Salesforce documentation: https://developer.salesforce.com/docs/atlas.en-us.api_asynch.meta/api_asynch/bulk_common_diff_two_versions.htm

## To test
- You can use [this Message](https://www.openfn.org/projects/wfwi-committed-giving-staging/messages/0651c6d3-9cbc-7c57-b2c7-df753008f6c7) and this job: https://github.com/OpenFn/women-for-women/blob/staging/jobs/committed/upsertCardMasterV2.js
- sandbox: see LP `WFW Full Sandbox Sept 2023`

**Comment by aleksa-krolls** [comment]
@mtuchi we are running into several timeouts and performance issues on WFWI... I think this will help. Can you take this on tomorrow and sync with @daissatou2 if any questions/ or if you need help testing?

**Comment by josephjclark** [comment]
Hi, so this is all a bit complicated (ain't it always)

First of there's the state of `jsforce`. V1 doesn't support the bulk2 API. v2 has been in beta for 3 years and does support bulk2. A month ago, `jsforce` [announced](url) that they're going to skip v2 and go straight to v3. It's not a good time to be updating that library.

It looks like manually creating our own bulk API operation is going to be quite a lot of engineering. Very doable, but annoying if `jsforce` is cooking their own solution.

What I'd really like to know, as quickly as possible and before we do any heavy engineering, is: does the bulk 2.0 API _actually_ help? I mean it could be a bug in our (or `jsforce`'s) code that's causing the problem.

I suggest we align development around this question, put out a "private" test adaptor version to test with WFWI, and watch the results for a few days (or however long we need). If we can prove bulk 2.0 helps, we'll focus on an engineering strategy for a general release with bulk 2.0 support (either our own or  built on `jsforce`).

Here's what I suggest:

* We cut a release of `salesforce` based on the beta version of `jsforce@2.0`, with support for the bulk2 API
* We engineer this release just enough that it can run the WFWI workflow. We take the least-effort solution wherever possible.
* We release a new major version with a `-alpha` suffix. We tag it `next`, not `latest`. So it'll be released as `salesforce@5.0.0-alpha`, tagged as `next`
* We add the release to platform and put WFWI on it. I don't know what the implications of this are for platform, I don't want other users to see this alpha version (let alone have it selected by default). We may need help on that.

Basically this alpha release is a thin, throwaway test version of the adaptor, used only by WFWI.

**Comment by mtuchi** [comment]
Hiya @josephjclark we are still facing this challenge of bulk job not closing 👇🏽 
![Screenshot 2024-03-15 at 2 35 19 PM](https://github.com/OpenFn/adaptors/assets/6592749/3b707fc5-1ce0-4512-ae4d-2008cc978627)

Can we make a plan for this today or early next week ?

**Comment by aleksa-krolls** [comment]
@mtuchi pls close if addressed!

**Comment by mtuchi** [comment]
@aleksa-krolls we should definitely pick this up when [jsforce v3 is released](https://github.com/jsforce/jsforce/issues/1360), I will put this on icebox for now. But ideally when v3 is out we should consider a major revamp of our salesforce adaptor. Couple of benefits of doing that are better docs for Salesforce API limitation, improved performance, clear pattern of what each function returns etc

**Comment by josephjclark** [comment]
I would actually close this issue down because we've solved the short-term problem.

Yes we should support bulk2 when jsforce3 is out - but that could take another year, and it's just clutter in our repo :(
